File: app/src/main/java/com/example/mvvmtest/viewmodel/ChatViewModel.java
package com.example.mvvmtest.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.mvvmtest.dagger.component.ApiController;
import com.example.mvvmtest.manager.FlechPreferences;
import com.example.mvvmtest.manager.JsonManager;
import com.example.mvvmtest.model.Message;
import com.example.mvvmtest.repository.ChatRepository;
import com.example.mvvmtest.util.Constant;
import com.github.nkzawa.emitter.Emitter;
import com.github.nkzawa.socketio.client.Socket;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;


public class ChatViewModel extends ViewModel {

    @Inject
    protected FlechPreferences flechPreferences;

    @Inject
    protected ChatRepository chatRepository;

    private int nickname;

    private MutableLiveData<List<Message>> mMessages;
    private MutableLiveData<Boolean> isTyping;

    public ChatViewModel() {
        ApiController.getAppComponent().inject(this);
    }

    public void init(int nickname) {
        this.nickname = nickname;
        mMessages = chatRepository.getMessages();
        isTyping = new MutableLiveData<>();
        isTyping.postValue(false);
    }


    @Override
    protected void onCleared() {
        super.onCleared();
    }

    public LiveData<List<Message>> getMessages() {
        return mMessages;
    }

    public LiveData<Boolean> getIsTyping() {
        return isTyping;
    }

    public Emitter.Listener onNewMessage = new Emitter.Listener() {
        @Override
        public void call(Object... args) {
            Log.d("GBC", "Entro on new message");
            Log.d("GBC", args[0].toString());
            processOnNewMessage((JSONObject) args[0]);
        }
    };

    public Emitter.Listener onTyping = new Emitter.Listener() {
        @Override
        public void call(Object... args) {

        }
    };

    private void processOnNewMessage(JSONObject data) {
        Message message = JsonManager.processNewMessage(data);
        if (message == null){
            return;
        }
        List<Message> list;
        if (mMessages != null) {
            list = mMessages.getValue();
            list.add(message);
        } else {
            list = new ArrayList<>();
            list.add(message);
        }

        mMessages.setValue(list);
    }

    private void processOnTyping(JSONObject data) {
        try {
            int sender = data.getInt("sender");
            int nickname = data.getInt("nickname");
            boolean isTyping = data.getBoolean("isTyping");


        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void sendMessage(Socket socket, Message message) {
        JSONObject object = JsonManager.createEmitSendMessage(message);
        socket.emit(Constant.SocketKey.FUNCTION_KEY, object);
    }

    public void sendTyping(Socket socket, int sender, int nickname) {
        socket.emit(Constant.SocketEvent.TYPING, sender, nickname);
    }


}

File: app/src/main/java/com/example/mvvmtest/viewmodel/ChatViewModel.kt
package com.example.mvvmtest.viewmodel


import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.mvvmtest.dagger.component.ApiController
import com.example.mvvmtest.manager.FlechPreferences
import com.example.mvvmtest.manager.JsonManager
import com.example.mvvmtest.model.Message
import com.example.mvvmtest.model.Typing
import com.example.mvvmtest.repository.ChatRepository
import com.example.mvvmtest.util.Constant
import com.github.nkzawa.emitter.Emitter
import com.github.nkzawa.socketio.client.Socket
import kotlinx.coroutines.*
import org.json.JSONObject
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext


class ChatViewModel(private val nickname: Int) : ViewModel() {

    @Inject
    lateinit var chatRepository: ChatRepository

    @Inject
    lateinit var flechPreferences: FlechPreferences

    val messagesLiveData: MutableLiveData<MutableList<Message>>
    val typingLiveData: MutableLiveData<Boolean>

    init {
        ApiController.getAppComponent().inject(this)
        messagesLiveData = MutableLiveData()
        typingLiveData = MutableLiveData()
    }

    //create a new Job
    private val parentJob = Job()
    //create a coroutine context with the job and the dispatcher
    private val coroutineContext: CoroutineContext
        get() = parentJob + Dispatchers.Default
    //create a coroutine scope with the coroutine context
    private val scope = CoroutineScope(coroutineContext)
    //live data that will be populated as news messages


    fun getMessages() {
        //launch the coroutine scope
        scope.launch {
            //get the messages
            val messages = chatRepository.getMessages(nickname)
            //post the value inside live data
            messagesLiveData.postValue(messages)
        }
    }

    fun sendMessage(socket: Socket, message: Message) {
        scope.launch {
            val jsonObject = JsonManager.createEmitSendMessage(message)
            socket.emit(Constant.SocketKey.FUNCTION_KEY, jsonObject)
        }
    }

    fun sendTyping(socket: Socket, typing: Typing) {
        scope.launch {
            val jsonObject = JsonManager.createEmitTyping(typing.sender, typing.nickname, typing.typing)
            socket.emit(Constant.SocketEvent.TYPING, jsonObject)
        }
    }

    val onNewMessageListener: Emitter.Listener = Emitter.Listener { args ->
        processOnNewMessage(args[0] as JSONObject)
    }

    val onTypingListener: Emitter.Listener = Emitter.Listener { args ->
        Log.d(Constant.TAG, "onTypingEvent ${args[0]}")
        processOnTyping(args[0] as JSONObject)
    }

    private fun processOnTyping(data: JSONObject) {
        val dataTyping = JsonManager.processTyping(data)
        if (!validateTyping(dataTyping as Typing)) return else typingLiveData.postValue(dataTyping.typing)
    }

    private fun processOnNewMessage(data: JSONObject) {
        val message = JsonManager.processNewMessage(data) ?: return
        if (!validateNewMessage(message)) {
            return
        } else {
            val newList = MutableList(1) { message }
            messagesLiveData.postValue(newList)
        }

    }

    private fun validateTyping(dataTyping: Typing): Boolean {
        return (Constant.SENDER == dataTyping.nickname) && (Constant.NICKNAME == dataTyping.sender)
    }

    private fun validateNewMessage(message: Message): Boolean {
        return ((Constant.SENDER == message.nickname) && (Constant.NICKNAME == message.sender) ||
                Constant.SENDER == message.sender && Constant.NICKNAME == message.nickname)
    }

    override fun onCleared() {
        super.onCleared()
        cancelRequest()
    }

    private fun cancelRequest() = coroutineContext.cancel()

}
